server/src/main/java/org/elasticsearch/gateway/ReplicaShardAllocator.java

```diff
@@ -311,36 +301,34 @@ public static Tuple<Decision, Map<String, NodeAllocationResult>> canBeAllocatedT
                 if (explain) {
                     madeDecision = decision;
                 } else {
-                    return Tuple.tuple(decision, null);
+                    return new PerNodeAllocationResult(decision, null);
                 }
             } else if (madeDecision.type() == Decision.Type.NO && decision.type() == Decision.Type.THROTTLE) {
                 madeDecision = decision;
             }
             if (explain) {
-                nodeDecisions.put(node.nodeId(), new NodeAllocationResult(node.node(), null, decision));
+                nodeDecisions.add(new NodeAllocationResult(node.node(), null, decision));
             }
         }
-        return Tuple.tuple(madeDecision, nodeDecisions);
+        return new PerNodeAllocationResult(madeDecision, nodeDecisions);
     }
 
+    public record PerNodeAllocationResult(Decision decision, List<NodeAllocationResult> nodes) {}
```

Comment:
Use a record instead of pair, since this is not part of private API anymore (already used in `SearchableSnapshotAllocator`)